### PR TITLE
[FrameworkBundle] fix removing "action" from route names

### LINKED ISSUE #40024: Controller actions ending with words like "action" are removing the "action" part from route name
**Symfony version(s) affected**: All

**Description**  
If you have in a controller some action like this one:

```php
/**
 * @Route("/create-transaction")
 */
public function createTransaction()
{
    // Code here
}
```

Then if you have to generate a route the action name will be `app_controller_createtrans`. If you want to have a proper action name like `app_controller_createtransaction` you should have the name of the action to `createTransactionAction`.

I know this is not exactly a bug and we can introduce a BC. It will be better for newcomers if the automatic name is the same as the method name except if it ends with an `Action` with capital `A`. WDYT? If someone decides that this should be done I want to propose a PR.

**Possible Solution**

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #40024
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by ro0NL** [comment]
should we change default names in a patch release?

**Comment by stof** [comment]
This will break all URL generation that uses the existing names. So it might indeed be dangerous in a patch release

**Comment by ro0NL** [comment]
also, from my experience, dereferencing virtual route names is pure pain :D

**Comment by nicolas-grekas** [comment]
> This will break all URL generation that uses the existing names. So it might indeed be dangerous in a patch release

Only when an action method doesn't end with `Action` nor `_action`, but still ends with `action`, in a case-sensitive way.

I'm fine doing this on 5.4...

**Comment by nicolas-grekas** [comment]
Now targetting 5.4

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Controller/AnnotatedController.php [hunk 1]
```diff
@@ -48,4 +48,12 @@ public function argumentWithoutDefaultWithRouteParamAndDefaultAction($value)
     {
         return new Response($value);
     }
+
+    /**
+     * @Route("/create-transaction")
+     */
+    public function createTransaction($value)
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Controller/AnnotatedController.php]: $value can be removed. it is irrelevant in this action

File: src/Symfony/Bundle/FrameworkBundle/Routing/AnnotatedRouteControllerLoader.php [hunk 2]
```diff
@@ -43,14 +43,12 @@ protected function configureRoute(Route $route, \ReflectionClass $class, \Reflec
...
-            '/action(_\d+)?$/',
-            '/__/',
-        ], [
-            '_',
-            '\\1',
-            '_',
-        ], parent::getDefaultRouteName($class, $method));
+        $name = preg_replace('/(bundle|controller)_/', '_', parent::getDefaultRouteName($class, $method));
+
+        if ('Action' === substr($method->name, -6) || '_action' === substr($method->name, -7)) {
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Routing/AnnotatedRouteControllerLoader.php]: IMO we should require the php 8 polyfill and use str_ends_with here. Ref. #41576

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Routing/AnnotatedRouteControllerLoader.php (hunk 2 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Routing/AnnotatedRouteControllerLoader.php]: I've opened a PR to do this in the whole codebase: #41973

**Comment by nicolas-grekas** [comment]
@Tobion comments addressed, thanks for the review!

**Review by Tobion — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @nicolas-grekas.

**Pull request merged**